Kubernetes: Provide properties to remote debugging

Sgitario opened the pull request:
This PR allows to easily configure the remote debugging and describes how to do this in the documentation.

I've also tried to expose another route to map the java agent port, but as intellij couldn't handle this port from localhost (it threw a handshake error), I decided to not over-complicate this guide and simple do the port-forward bit (see my changes in documentation).

I only did these changes for OpenShift and Kubernetes, not Knative, but let me know if you think it could fit in Knative as well. 

Fix https://github.com/quarkusio/quarkus/issues/17581
Fix https://github.com/quarkusio/quarkus/issues/23765

Sgitario commented:
/cc @geoand @iocanel @edeandrea @cmoulliard

geoand reviewed (commented):
Nice idea!

cmoulliard commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -1214,6 +1222,21 @@ In other words the extension will use whatever cluster `kubectl` uses. The same
 
+=== Remote Debugging
+
+To remotely debug applications that are running on a cloud environment, we need to deploy the application as described in the previous section and additionally appending the property `quarkus.kubernetes.remote-debug.enabled=true`. This property will automatically configure the Java application to append the java agent configuration (for example: `-agentlib:jdwp=transport=dt_socket,server=y,suspen ...
```
`cloud environment` --> `kubernetes environment`
`and additionally appending the property "quarkus.kubernetes.remote-debug.enabled=true"` --> `and add as new property: "quarkus.kubernetes.remote-debug.enabled=true"`
`and also the service resource to listen the java agent port. ` --> `and also the service resource to listen using the java agent port.`

Sgitario replied:
updated

cmoulliard commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -1214,6 +1222,21 @@ In other words the extension will use whatever cluster `kubectl` uses. The same
+
+To remotely debug applications that are running on a cloud environment, we need to deploy the application as described in the previous section and additionally appending the property `quarkus.kubernetes.remote-debug.enabled=true`. This property will automatically configure the Java application to append the java agent configuration (for example: `-agentlib:jdwp=transport=dt_socket,server=y,suspen ...
+
+After your application has been deployed with the debug enabled, next you need to port-forward the service java agent port:app-name:
```
`... you need to to port-forward the service java agent port:app-name:` --> `... you need to tunnel the traffic from your local host machine to the specified port of the java agent`

Sgitario replied:
changed

cmoulliard commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -1214,6 +1222,21 @@ In other words the extension will use whatever cluster `kubectl` uses. The same
+kubectl port-forward svc/<application name> 5005:5005
+----
+
+Using this command, you'll be forwaring all the TCP traffic from `localhost:5005` to the service name with port 5005 which is the one that the java agent uses by default for remote debugging. You can also configure another java agent port using the property `quarkus.kubernetes.remote-debug.address-port`.
```
`you'll be forwaring` --> `you'll forward the traffic from the "localhost:5005" to the kubernetes service running the java agent using the port "5005"`

Sgitario replied:
changed

cmoulliard commented on docs/src/main/asciidoc/deploying-to-kubernetes.adoc:
```diff
@@ -1214,6 +1222,21 @@ In other words the extension will use whatever cluster `kubectl` uses. The same
+
+Using this command, you'll be forwaring all the TCP traffic from `localhost:5005` to the service name with port 5005 which is the one that the java agent uses by default for remote debugging. You can also configure another java agent port using the property `quarkus.kubernetes.remote-debug.address-port`.
+
+Finally, all you need to do is to configure your favorite IDE to attach the java agent process that is forwarded to `localhost:5005` and start debugging your application. For example, in IntelliJ IDEA, you can follow https://www.jetbrains.com/help/idea/tutorial-remote-debug.html:[this tutorial] to debug remote applications.
```
`and start debugging your application` --> `... and start to debug your application.`

Sgitario replied:
changed

edeandrea commented:
Will setting the `JAVA_TOOL_OPTIONS` environment variable have any effect if the `command` on the `DeploymentConfig` doesn't reference it? I suppose it will according to https://docs.oracle.com/en/java/javase/11/troubleshoot/environment-variables-and-system-properties.html#GUID-BE6E7B7F-A4BE-45C0-9078-AA8A66754B97

I still think the `DeploymentConfig` [in my comment](https://github.com/quarkusio/quarkus/issues/23765#issuecomment-1048785936) is wrong though and needs to be fixed. The `command` is

```yaml
command:
  - java
  - -Dquarkus.http.host=0.0.0.0
  - -Djava.util.logging.manager=org.jboss.logmanager.LogManager
  - -jar
  - /deployments/quarkus-run.jar
```

Shouldn't that be fixed as well?

Sgitario commented:
> Will setting the `JAVA_TOOL_OPTIONS` environment variable have any effect if the `command` on the `DeploymentConfig` doesn't reference it? I suppose it will according to https://docs.oracle.com/en/java/javase/11/troubleshoot/environment-variables-and-system-properties.html#GUID-BE6E7B7F-A4BE-45C0-9078-AA8A66754B97
> 

Yes, using the changes in this PR, users won't need to deal with how to provide the JVM arguments any longer. Internally, it will use the `JAVA_TOOL_OPTIONS` env variable which works fine to enable remote debugging.

About [your comment](https://github.com/quarkusio/quarkus/issues/23765#issuecomment-1048785936), you mentioned several things:
- `the imagePullPolicy doesn't seem to be respected`. Indeed. You reported it in a separated issue and it will be fixed as part of this PR: https://github.com/quarkusio/quarkus/pull/24068
- wrong generated manifests when providing the `JAVA_DEBUG` env var:

Extracted from the comment:

```In https://github.com/quarkusio/quarkus-super-heroes/tree/main/rest-villains If I run

./mvnw clean package -DskipTests \
  -Dquarkus.kubernetes.deployment-target=openshift \
  -Dquarkus.container-image.builder=openshift \
  -Dquarkus.openshift.resources.limits.memory=768Mi \
  -Dquarkus.openshift.resources.requests.memory=256Mi \
  -Dquarkus.openshift.env.vars.JAVA_DEBUG=true

and then inspect the resulting target/kubernetes/openshift.yml I see that the app's DeploymentConfig looks like the following and the command is overridden.

In this scenario, it is almost impossible to turn on remote debug unless I then go in and manually tweak the resulting DeploymentConfig. The DeploymentConfig already references the image (which is already set up for remote debug), so why the need to also specify a command?
```

The command is auto-generated by the OpenShift extension and you would only specify one at your own risk. 
In order to provide JVM arguments, as you did in previous comments, you should use `quarkus.openshift.jvm-arguments`, what seems wrong to me is that the Quarkus extension is splitting the arguments by comma. I've opened a separate ticket for fixing this: https://github.com/quarkusio/quarkus/issues/24136

Sgitario committed:
Kubernetes: Provide properties to remote debugging

This PR allows to easily configure the remote debugging and describes how to do this in the documentation.

I've also tried to expose another route to map the java agent port, but as intellij could handle this port from localhost (it threw a handshake error), I decided to not over-complicate this guide and simple do the port-forward bit (see my changes in documentation).

Fix https://github.com/quarkusio/quarkus/issues/17581
Fix https://github.com/quarkusio/quarkus/issues/23765

edeandrea commented:
> The command is auto-generated by the OpenShift extension and you would only specify one at your own risk.

But why does the `DeploymentConfig` generate a `command` section in the first place? To me that seems wrong. Why does the `DeploymentConfig` contain a `command`? It doesn't need to since the image itself contains a `CMD`/`ENTRYPOINT`.

If I were to `docker run` the image I wouldn't need to specify a command to run. It would just run the image.

iocanel commented:
> > > The command is auto-generated by the OpenShift extension and you would only specify one at your own risk.
> > 
> > 
> > But why does the `DeploymentConfig` generate a `command` section in the first place? To me that seems wrong. Why does the `DeploymentConfig` contain a `command`? It doesn't need to since the image itself contains a `CMD`/`ENTRYPOINT`.
> > If I were to `docker run` the image I wouldn't need to specify a command to run. It would just run the image.
> 
> I guess this decision was made to be able to use as many images as possible since some images might have a not Quarkus-compatible command, but I will add @iocanel to the thread, so he can clarify this point.

@Sgitario @edeandrea: I don't recall what was the initial idea. I do recall though that in other places we assume that it's the images responsibility to set the command unless the user explicitly specifies one. Eitherway, I think that at some point we can change that.

iocanel reviewed (commented):
I am pretty sure this will need to be rebased.

iocanel reviewed (approved):
The PR looks, good to me. But I think that it needs rebase.

edeandrea commented:
> I do recall though that in other places we assume that it's the images responsibility to set the command unless the user explicitly specifies one.

I would completely agree with that @iocanel. In most cases, if I'm running some image I would assume that the command to run it is already baked into the image.
